src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -0,0 +1,1036 @@
+        }
+
+        public void CheckSetResetConnectionState(uint error, CallbackType callbackType)
+        {
+            // Should only be called for MARS - that is the only time we need to take
+            // the ResetConnection lock!
+
+            // It was raised in a security review by Microsoft questioning whether
+            // we need to actually process the resulting packet (sp_reset ack or error) to know if the
+            // reset actually succeeded.  There was a concern that if the reset failed and we proceeded
+            // there might be a security issue present.  We have been assured by the server that if
+            // sp_reset fails, they guarantee they will kill the resulting connection.  So - it is
+            // safe for us to simply receive the packet and then consume the pre-login later.
+
+            Debug.Assert(_parser.MARSOn, "Should not be calling CheckSetResetConnectionState on non MARS connection");
+
+            if (_fResetEventOwned)
+            {
+                if (callbackType == CallbackType.Read && TdsEnums.SNI_SUCCESS == error)
+                {
+                    // RESET SUCCEEDED!
+                    // If we are on read callback and no error occurred (and we own reset event) -
+                    // then we sent the sp_reset_connection and so we need to reset sp_reset_connection
+                    // flag to false, and then release the ResetEvent.
+                    _parser._fResetConnection = false;
+                    _fResetConnectionSent = false;
+                    _fResetEventOwned = !_parser._resetConnectionEvent.Set();
+                    Debug.Assert(!_fResetEventOwned, "Invalid AutoResetEvent state!");
+                }
+
+                if (TdsEnums.SNI_SUCCESS != error)
+                {
+                    // RESET FAILED!
+
+                    // If write or read failed with reset, we need to clear event but not mark connection
+                    // as reset.
+                    _fResetConnectionSent = false;
+                    _fResetEventOwned = !_parser._resetConnectionEvent.Set();
+                    Debug.Assert(!_fResetEventOwned, "Invalid AutoResetEvent state!");
+                }
+            }
+        }
+
+        internal void CloseSession()
+        {
+            ResetCancelAndProcessAttention();
+#if DEBUG
+            InvalidateDebugOnlyCopyOfSniContext();
+#endif
+            Parser.PutSession(this);
+        }
+
+        internal bool Deactivate()
+        {
+            bool goodForReuse = false;
+
+            try
+            {
+                TdsParserState state = Parser.State;
+                if (state != TdsParserState.Broken && state != TdsParserState.Closed)
+                {
+                    if (HasPendingData)
+                    {
+                        Parser.DrainData(this); // This may throw - taking us to catch block.c
+                    }
+
+                    if (HasOpenResult)
+                    {
```

Comment:
```suggestion
                    { // SQL BU DT 383773 - need to decrement openResultCount for all pending operations.
```
keeping the comment from netfx